### PR TITLE
BAH 2962 | EventListener to send SMS to the user

### PR DESCRIPTION
**Pull request by anubhavBeehyv**
- Added an event listener to listen to the events coming up from bahmni-events module
- Events on which sms is send as for now are patient creation,appointment booking(recurring,regular).
- Added an appointment reminder sms feature in the same (working as a part of scheduler).

### CODE REVIEW

File: api/pom.xml [hunk 1]
```diff
@@ -141,6 +141,11 @@
             <version>${powerMockVersion}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.5.14</version>
```
Comment (angshu) [inline api/pom.xml]: scope should be provided

File: api/src/main/java/org/bahmni/module/communication/service/impl/CommunicationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			SMSRequest smsRequest = new SMSRequest();
+			smsRequest.setPhoneNumber(phoneNumber);
+			smsRequest.setMessage(message);
+			ObjectMapper objMapper = new ObjectMapper();
+			String jsonObject = objMapper.writeValueAsString(smsRequest);
+			StringEntity params = new StringEntity(jsonObject);
+			String smsUrl = Context.getAdministrationService().getGlobalProperty("bahmni.sms.url", SMS_URI);
+			HttpPost request = new HttpPost(Context.getMessageSourceService().getMessage(smsUrl, null, new Locale("en")));
+			request.addHeader("content-type", "application/json");
+			String tokenFilePath = new File("/tmp/sms-tokens", SMS_TOKEN_KEY + "-token.txt").getAbsolutePath();
```
Comment (angshu) [inline api/src/main/java/org/bahmni/module/communication/service/impl/CommunicationServiceImpl.java]: should be reading from a file available under OMRS application data directory

File: api/src/main/java/org/bahmni/module/communication/service/impl/CommunicationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			String token = messagingUtility.getSMSTokenFromTokenFile(tokenFilePath);
+			
+			if (token == null) {
+				throw new RuntimeException("Token not found in the token file: " + tokenFilePath);
+			}
+			
+			request.addHeader("Authorization", "Bearer " + token);
+			request.setEntity(params);
+			CloseableHttpClient httpClient = HttpClients.createDefault();
```
Comment (angshu) [inline api/src/main/java/org/bahmni/module/communication/service/impl/CommunicationServiceImpl.java:124]: maybe better done leveraging closeable interface. Check https://www.baeldung.com/apache-httpclient-vs-closeablehttpclient

File: api/src/main/resources/liquibase.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
Comment (angshu) [inline api/src/main/resources/liquibase.xml]: do we need this? if not, remove

File: api/src/main/resources/messages.properties [hunk 5]
```diff
@@ -0,0 +1,9 @@
+sms.appointmentReminderSMSTemplate=This is a reminder for upcoming Appointment for {patientName}:{identifier}\non Date:{date} for {service} at {facilityname}.
+sms.helpdeskTemplate=For any queries call us on {helpdeskNumber}.
+sms.providersTemplate=Your appointment is booked with {providerName}.
+
+sms.appointmentBookingSMSTemplate=Appointment for {patientName}, {identifier} on {date} for {service} is booked at {facilityname}.
+sms.teleconsultationLinkTemplate=As you have chosen to book a teleconsultation-appointment click on the link {teleconsultationLink} on {date} to join the consultation.
+sms.recurringAppointmentTemplate=A recurring appointment for {patientName}, {identifier} for {service} is booked for every {recurringPeriod} {recurringType} {recurringDays} at {facilityname}. Your next appointment is at {date}.
+
+bahmni.sms.url=http://sms-service:8080/notification/sms
```
Comment (angshu) [inline api/src/main/resources/messages.properties]: Why is the sms server url here? it should be in global property. If the property is not found - log error.

**Review by angshu — CHANGES_REQUESTED** [review]

**Review by angshu — APPROVED** [review]

**Comment by angshu** [comment]
You have lowered the ratio, which is ok for now but please create a card on JIRA to improve test coverage for this module.

**Comment by anubhavBeehyv** [comment]
> You have lowered the ratio, which is ok for now but please create a card on JIRA to improve test coverage for this module.

Sure @angshu

**Pull request merged**